samza-core/src/main/java/org/apache/samza/execution/JobNodeConfigureGenerator.java

```diff
@@ -0,0 +1,340 @@
+      });
+  }
+
+  /**
+   * Serializes the {@link Serde} instances for operators, adds them to the provided config, and
+   * sets the serde configuration for the input/output/intermediate streams appropriately.
+   *
+   * We try to preserve the number of Serde instances before and after serialization. However we don't
+   * guarantee that references shared between these serdes instances (e.g. an Jackson ObjectMapper shared
+   * between two json serdes) are shared after deserialization too.
+   *
+   * Ideally all the user defined objects in the application should be serialized and de-serialized in one pass
+   * from the same output/input stream so that we can maintain reference sharing relationships.
+   *
+   * @param configs the configs to add serialized serde instances and stream serde configs to
+   */
+  private void configureSerdes(Map<String, String> configs, Map<String, StreamEdge> inEdges, Map<String, StreamEdge> outEdges,
+      List<StoreDescriptor> stores, JobNode jobNode) {
+    // collect all key and msg serde instances for streams
+    Map<String, Serde> streamKeySerdes = new HashMap<>();
+    Map<String, Serde> streamMsgSerdes = new HashMap<>();
+    inEdges.keySet().forEach(streamId ->
+        addSerde(jobNode.getInputSerde(streamId), streamId, streamKeySerdes, streamMsgSerdes));
+    outEdges.keySet().forEach(streamId ->
+        addSerde(jobNode.getOutputSerde(streamId), streamId, streamKeySerdes, streamMsgSerdes));
+
+    Map<String, Serde> storeKeySerdes = new HashMap<>();
+    Map<String, Serde> storeMsgSerdes = new HashMap<>();
+    stores.forEach(storeDescriptor -> {
+        storeKeySerdes.put(storeDescriptor.getStoreName(), storeDescriptor.getKeySerde());
+        storeMsgSerdes.put(storeDescriptor.getStoreName(), storeDescriptor.getMsgSerde());
+      });
+
+    // for each unique stream or store serde instance, generate a unique name and serialize to config
+    HashSet<Serde> serdes = new HashSet<>(streamKeySerdes.values());
+    serdes.addAll(streamMsgSerdes.values());
+    serdes.addAll(storeKeySerdes.values());
+    serdes.addAll(storeMsgSerdes.values());
+    SerializableSerde<Serde> serializableSerde = new SerializableSerde<>();
+    Base64.Encoder base64Encoder = Base64.getEncoder();
+    Map<Serde, String> serdeUUIDs = new HashMap<>();
+    serdes.forEach(serde -> {
+        String serdeName = serdeUUIDs.computeIfAbsent(serde,
+            s -> serde.getClass().getSimpleName() + "-" + UUID.randomUUID().toString());
+        configs.putIfAbsent(String.format(SerializerConfig.SERDE_SERIALIZED_INSTANCE(), serdeName),
+            base64Encoder.encodeToString(serializableSerde.toBytes(serde)));
+      });
+
+    // set key and msg serdes for streams to the serde names generated above
+    streamKeySerdes.forEach((streamId, serde) -> {
+        String streamIdPrefix = String.format(StreamConfig.STREAM_ID_PREFIX(), streamId);
+        String keySerdeConfigKey = streamIdPrefix + StreamConfig.KEY_SERDE();
```

Comment:
The original goal was to serialize all serde instances in one pass to preserve the number of serde instances before and after deserialization. However Tables already break that guarantee since they serialize their Serde separately, so we can consider moving this logic to StreamDescriptor.

Reply:
@prateekm I like the idea, that would actually simplify the serde configuration generation a lot. When we also have Input/OutputDescriptor for intermediate streams, we don't even need this centralized serde generation method. I will work on that.

Reply:
Actually, it might be nicer to keep a per JobNode serde map that are used by all serde configure generation for all streams/stores/tables. Let me work on that.